Document TIMESYNC protocol (#433)
mavlink/mavlink-devguide: en/services/timesync_v1.md

# Time Synchronization Protocol v1

This protocol is used to estimate the time offsets between system clocks on two MAVLink systems, in a two-system MAVLink network.

> This is useful because it allows ....

Author: What are the use cases? I understand this gives you the difference in system clock, but when might you need to know that?

Author: @peterbarker So why do you need this?

Author: It's generally nice for people who are learning about a system to understand what it is for/why they should care.

The protocol uses just one message [TIMESYNC](#TIMESYNC), which has two `int64_t` fields, `tc1` and `ts1`. 

> The message is sent out with a timestamp in `ts1` (current unix timestamp *1000).

Reviewer: unix timestamps are in seconds.  ArduPilot uses nanoseconds in this field, so for AP this should be
```suggestion
The message is sent out with a timestamp in `ts1` (current unix timestamp *1e9).
```

Why is it nanoseconds?  Good question.  I may have found something out there that indicated that was the relevant scale.

Author: Thanks @peterbarker . 

@julianoes Peter is right that that unix timestamp units are in seconds. The TIMESYNC message doesn't specify a unit (and should).  I think that PX4 essentially has the same interpretation - from the RTPS code where `hrt_absolute_time` is in us - this is the same right?

```
@(topic)_data.tc1 = hrt_absolute_time() * 1000ULL;
```

Reviewer 2: Yes, from what I can see PX4 also uses nanoseconds, phew!

> The message is sent out with a nanosecond timestamp in `ts1` (current unix timestamp *1e9).

Author: ```suggestion
The message is sent out with a nanosecond timestamp in `ts1` (current unix timestamp * 1e9).
```

> The message is sent out with a nanosecond timestamp in `ts1` (current unix timestamp * 1e9).

Reviewer 3: Do existing implementations use UNIX timestamps? It appears not.

Author: What makes you think they aren't?

Reviewer 3: [The PX4 implementation](https://github.com/PX4/PX4-Autopilot/blob/0ea347a5c9f3f419a01f6b8e7f7952e7455b9aa7/src/modules/mavlink/mavlink_timesync.cpp#L52) uses `hrt_absolute_time()`. I'm under the impression that `hrt_absolute_time()` returns a timestamp from a hardware timer, not a unix timestamp.

Author: @julianoes do you know if @obicons is correct ? I.e. I am absolutely sure the method returns a timestamp from the hardware timer but I am not sure whether or not there is anything else that then synchronizes that to an absolute time. I highly suspect that even if it is possible, you can't rely on it.

As I understand it timesync is used mostly to provide accurate relative time sync - ie it doesn't matter that they aren't at the same time as long as the offset in values provided by both is known. 
- Is that correct?
- Are there use cases where the values _need_ to be a timestamp against a known reference like unix epoch?

If not, I'd suggest we change the message to make it seconds since boot when I add the extensions for sysid/component id.

Reviewer 3: Other messages allow timestamps in both formats (i.e., epoch and system boot.) Perhaps allow either in this message, and update the documentation?

Author: Fair enough. The way to progress this is to test it in a message update: https://github.com/mavlink/mavlink/pull/1878

Come back here when we work out the appetite for change.

This is mirrored back by the remote system, along with its own timestamp.
The first system can then determine the round-trip time, and estimate the timestamp offset.

> This sequence is run multiple times to remove the transient channel and processor usage from the offset calculation.

Author: Note that originally I thought this was because the timesync would drift with time. But actually it is because the round trip time, and time spent on each leg is not deterministic. So we need to average out the values and filter outliers.

The protocol only works when there are only two systems on the MAVLink network.
This is because [TIMESYNC](#TIMESYNC) is broadcast (cannot be addressed to a single node), so multiple components on the network may respond.

> A recipient has no way to reliably determine if the timestamp in `ts1` is the one it sent out, or is a reply to another system.

Reviewer: ArduPilot encodes its own system ID in the bottom-most bits of the nano-second field.  I suggest that's a reasonable pattern other GCSs could adopt.

```
    // we add in our own system id try to ensure we only consider
    // responses to our own timesync request messages
    return AP_HAL::micros64()*1000LL + mavlink_system.sysid;
```

Author: @peterbarker Thanks very much.  I'd have preferred we did it the "mavlink way" and had system/component ids in the message. Mostly because then you wouldn't be requiring every system on the network to respond, even if you aren't very interested.

Other than the unnecessary traffic on a very busy network this is great because the timestamp is mirrored back and used by the same system - no compatibility issues. @julianoes Are you OK with this?

Reviewer 2: Interesting. PX4 also uses nanoseconds in the time field, so I think we could adopt it.

So that's essentially the same as the `target_system` and `target_component` fields, or the source `sysid`, `compid`?

If it's the same as `target_system` and `target_component` I would strongly suggest that we add that in as an extension.

Or would that create problems that I'm missing?

Author: @julianoes ArduPilot just copies in the system_id (not the component id, so you can't sync to a component). The value copied in is the sender id. At the end the value is sent back by the recipient unchanged - so when the original sender gets it back it can tell it is the timestamp it sent out.

So 
1. not the target_system/component being sent. 
2. not quite as flexible as if we did that - since it is only system to system.
3. But less disruptive. Because values are just mirrored back you don't need to think about compatibility with legacy systems or anything really.

Reviewer 2: So what if you have a different component ID?

I think we should tolerate the existing implementation and note it down but extend the message in a way to properly support system and component IDs.

Author: @julianoes Thanks. I have proposed the text in https://github.com/mavlink/mavlink-devguide/pull/433/files#r936287659 to reflect the situation now in "timesync v1". This includes the system id.

If we were to add system/component ID then we'd create a v2 version in which you would use the fields as proposed.

I can't add that unless someone is willing to implement it - no stakeholder.

Why should we do it? 
- It is more mavlink, and therefore less of a hack and confusing for all concerned.
- This would work for components too. Is it likely you'll want to have multiple components syncing in one system?
- Reduced bandwidth. You could sync to just one component (say) that you are interested in, and avoid responses from every other system on the network.

I lean towards doing it as "the right thing", but against because I don't think ArduPilot will update.  Unless we see the enough value for the component use case and reduced bandwith, I'm against complicating our different implementations further.

Reviewer 2: > If we were to add system/component ID then we'd create a v2 version in which you would use the fields as proposed

Why not extensions?

Author: Sorry, I mean a "v2 version of the protocol". I'd be happy enough with the extensions for the messages.

What I'm worried about here is microservices management. e.g. how a system knows what protocols are supported, what if the extensions are implemented on a sender system but not a reciever etc etc. To me the clearest way to handle that is to call this a new protocol. 

Perhaps I am over thinking it - but it's easy to add two fields as an extension and miss all the other "questions". And as a reader, it is much easier to deprecate the ardupilot way of doing things in a separate doc than mix it in just one doc.

Reviewer 2: Right. The problem I see is that new messages are very slow to get picked up and put the burden on the implementer that now need to support both ways.

> The protocol only works when there are only two systems on the MAVLink network.
> This is because [TIMESYNC](#TIMESYNC) is broadcast (cannot be addressed to a single node), so multiple components on the network may respond.
> A recipient has no way to reliably determine if the timestamp in `ts1` is the one it sent out, or is a reply to another system.

Author: ```suggestion
[TIMESYNC](#TIMESYNC) is a broadcast message, so all components on the network will receive all `TIMESYNC` requests, and all components that support the protocol will send back `TIMESYNC` responses.
If there are multiple systems on the network, a system that is waiting on a `TIMESYNC` response cannot be _certain_ that a response to its request and not to that of another components.

A component can reduce the chance of clashes by appending its system ID to the bottom-most bits of the timestamp sent in the request, and only calculating the sync on responses that include this id.
This is approach is recommended, and should have no side effects, because the timestamp is simply mirrored by the pinged system.
```

## Message/Enum Summary

Message | Description
-- | --
<a id="TIMESYNC"></a>[TIMESYNC](../messages/common.md#TIMESYNC) | Time synchronisation message.

## Sequences

[![Mermaid sequence: Time sync](https://mermaid.ink/img/pako:eNqFUT1vwjAQ_SsnT0UKKMnoCpaWoQMsdKmU5WpfitXYTu0LFCH-e52EQlWh1tP5-X1ZdxTKaxJSRProyCl6NPgW0N5XDtJpMbBRpkXHsMJdcRMtR3TtmcDvKAxY1tPlIILQe0eWwKqY5xlwLObDAxubcLTt6NBj08Wil0t4flotNy_rB7j7oXJ-D59Q5Hk--Tu0TKFtcxgjR-ASdqvA7NKgnI4Vit8VrtlXg1kaJv99f4sRgu-cBg6mHTIB0y0lG4tMGnxdR2J4Jd4TOVCNV-8xVRqdRSYsBYtGpz0de6wSvCVLlZBp1FRj13AlKndK1K7VyXOpDfsgZI1NpExgx35zcEpIDh19k867PrNOX9YgqSA)](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNqFUT1vwjAQ_SsnT0UKKMnoCpaWoQMsdKmU5WpfitXYTu0LFCH-e52EQlWh1tP5-X1ZdxTKaxJSRProyCl6NPgW0N5XDtJpMbBRpkXHsMJdcRMtR3TtmcDvKAxY1tPlIILQe0eWwKqY5xlwLObDAxubcLTt6NBj08Wil0t4flotNy_rB7j7oXJ-D59Q5Hk--Tu0TKFtcxgjR-ASdqvA7NKgnI4Vit8VrtlXg1kaJv99f4sRgu-cBg6mHTIB0y0lG4tMGnxdR2J4Jd4TOVCNV-8xVRqdRSYsBYtGpz0de6wSvCVLlZBp1FRj13AlKndK1K7VyXOpDfsgZI1NpExgx35zcEpIDh19k867PrNOX9YgqSA)

The sequence is:

> 1. The first system, which needs time synchronisation, broadcasts `TIMESYNC` with `tc1 = 0`, and `ts1 = timestamp "now" * 1000`.
> 1. The second system receives the message with `TIMESYNC.tc1 == 0`, and hence knows it is a`TIMESYNC` request from another system.
>    It responds with the first system's timestamp mirrored back in `ts1`, and its own timestamp in `tc1` (`timestamp "now" * 1000`).

Author: ```suggestion
1. The first system, which needs time synchronisation, broadcasts `TIMESYNC` with `tc1 = 0`, and `ts1 = timestamp "now"  * 1e9`.
1. The second system receives the message with `TIMESYNC.tc1 == 0`, and hence knows it is a`TIMESYNC` request from another system.
   It responds with the first system's timestamp mirrored back in `ts1`, and its own timestamp in `tc1` (`timestamp "now" * 1e9`).
```
